lib/core/test/unit/Cardano/Pool/DB/Properties.hs

```diff
@@ -844,6 +851,90 @@ prop_rollbackRetirement DBLayer{..} certificates =
         , ii <- [0 .. 3]
         ]
 
+-- When we remove pools, check that:
+--
+-- 1. We only remove data relating to the specified pools.
+-- 2. We do not remove data relating to other pools.
+--
+prop_removePools
+    :: DBLayer IO
+    -> [PoolCertificate]
+    -> Property
+prop_removePools
+    DBLayer {..} certificates =
+        monadicIO (setup >> prop)
+  where
+    setup = run $ atomically cleanDB
+
+    prop = do
+        -- Firstly, publish an arbitrary set of pool certificates:
+        run $ atomically $ do
+            mapM_ (uncurry putCertificate) certificatePublications
+        -- Next, read the latest certificates for all pools:
+        poolIdsWithRegCertsAtStart <- run poolIdsWithRegCerts
+        poolIdsWithRetCertsAtStart <- run poolIdsWithRetCerts
+        -- Next, remove a subset of the pools:
+        run $ atomically $ removePools $ Set.toList poolsToRemove
+        -- Finally, see which certificates remain:
+        poolIdsWithRegCertsAtEnd <- run poolIdsWithRegCerts
+        poolIdsWithRetCertsAtEnd <- run poolIdsWithRetCerts
+        monitor $ counterexample $ T.unpack $ T.unlines
```

Comment:
Is this amount of set-arithmetic really needed / worth it?

Though I guess the error messages are nicer than if the counterexample was just the specific element that caused the failure.

Reply:
> Is this amount of set-arithmetic really needed / worth it?

I think you're right that there is some redundancy in the set arithmetic checks at the end. :+1:  But I think this redundancy is okay, as it can help to make failures a bit clearer when debugging.